js/__internal/grids/grid_core/keyboard_navigation/module.ts

```diff
@@ -2107,28 +2125,30 @@ export class KeyboardNavigationController extends modules.ViewController {
     return args;
   }
 
-  _fireFocusedRowChanged($rowElement) {
-    let row;
-
+  _fireFocusedRowChanged(): void {
+    const focusedRowEnabled = this.option('focusedRowEnabled');
     const focusedRowKey = this.option('focusedRowKey');
+    const focusedRowIndex = this._focusController?.getFocusedRowIndexByKey(focusedRowKey);
 
-    const focusController = this.getController('focus');
-    const focusedRowIndex = focusController?.getFocusedRowIndexByKey(focusedRowKey);
-
-    if (this.option('focusedRowEnabled')) {
-      if (focusedRowIndex >= 0) {
-        const dataController = this._dataController;
-        row = focusedRowIndex >= 0
-          && dataController.getVisibleRows()[
-            focusedRowIndex - dataController.getRowIndexOffset()
-          ];
-      }
-      this.executeAction('onFocusedRowChanged', {
-        rowElement: $rowElement,
-        rowIndex: focusedRowIndex,
-        row,
-      });
+    if (!focusedRowEnabled || (isDefined(focusedRowKey) && focusedRowIndex < 0)) {
+      return;
     }
+
+    this._memoFireFocusedRowChanged(focusedRowKey, focusedRowIndex);
+  }
+
+  _memoFireFocusedRowChanged(focusedRowKey: unknown, focusedRowIndex: number): void {
```

Comment:
Minor:

```suggestion
  _memoFireFocusedRowChanged(focusedRowKey: Key, focusedRowIndex: number): void {
```

Let's add `type Key = unknown` in module_types near to UserData. Maybe later we'll replace it with generic type or with nominal type

Reply:
Discussed.

We decided to create a new type ```RowKey = unknown``` in ```module_types.ts``` at grid-core.
And move the ```UserData``` to it too.